ZEngine/ZEngine/Core/Maths/Matrix.h

```diff
@@ -0,0 +1,485 @@
+#include <Helpers/MemoryOperations.h>
+#include <cstddef>
+#include "Vec.h"
+
+using namespace ZEngine::Helpers;
+namespace ZEngine::Core::Maths
+{
+    template <typename T, size_t R, size_t C, typename = std::enable_if_t<std::is_arithmetic_v<T> && (R >= 1) && (C >= 1)>>
+    struct Matrix
+    {
+        T  m_data[R * C];
+
+        T& operator()(size_t row, size_t col)
+        {
+            ZENGINE_VALIDATE_ASSERT(row < R && col < C, "Index out of range");
+            return m_data[row * C + col];
+        }
+
+        const T& operator()(size_t row, size_t col) const
+        {
+            ZENGINE_VALIDATE_ASSERT(row < R && col < C, "Index out of range");
+            return m_data[row * C + col];
+        }
+
+        Matrix<T, R, C> operator+(Matrix<T, R, C>& other)
+        {
+            Matrix<T, R, C> result{};
+            for (size_t j = 0; j < C; ++j)
+            {
+                for (size_t i = 0; i < R; i++)
+                {
+                    result(i, j) = other(i, j) + (*this)(i, j);
+                }
+            }
+            return result;
+        }
+
+        Matrix<T, R, C> operator-(Matrix<T, R, C>& other)
+        {
+            Matrix<T, R, C> result{};
+            for (size_t j = 0; j < C; ++j)
+            {
+                for (size_t i = 0; i < R; i++)
+                {
+                    result(i, j) = (*this)(i, j) - other(i, j);
+                }
+            }
+            return result;
+        }
+    };
+
+    template <typename T>
+    struct Mat2 : public Matrix<T, 2, 2>
+    {
+
+        Mat2()
+        {
+            secure_memset(this->m_data, 0, sizeof(T) * 4, sizeof(T) * 4);
+        }
+
+        Mat2(T m00, T m01, T m10, T m11)
+        {
+            this->m_data[0] = m00;
+            this->m_data[1] = m01;
+            this->m_data[2] = m10;
+            this->m_data[3] = m11;
+        }
+
+        Mat2(const Matrix<T, 2, 2>& other)
+        {
+            for (size_t i = 0; i < 4; ++i)
+                this->m_data[i] = other.m_data[i];
```

Comment:
instead of looping , we can directly index data e.g `m_data[0][0] = other.m_data[0][0]`